adding docs on how to access data in pipeline (#9611)
opensearch-project/documentation-website: _ingest-pipelines/accessing-data.md

---
layout: default
title: Accessing data in pipeline

> redirect_from:

Reviewer: Redirects are only necessary for existing pages that have moved. If it's a brand new page (and is not an index page), redirect is not necessary. 
```suggestion
```

  - /opensearch/rest-api/ingest-apis/accessing-data/
  - /api-reference/ingest-apis/accessing-data/
nav_order: 20
---

# Accessing data in pipelines

> In ingest pipelines, you access the data of incoming documents using the `ctx` object. This object represents the document being processed and allows you to read, modify, or enrich its fields. Processors in a pipeline have read and write access to both the `_source` fields of a document and its metadata fields.

Reviewer: ```suggestion
In ingest pipelines, you access the document data using the `ctx` object. This object represents the processed document and allows you to read, modify, or enrich the document fields. Processors in a pipeline have read and write access to both the `_source` fields of a document and its metadata fields.
```

## Accessing document fields

The `ctx` object exposes all document fields. You can access them directly using dot notation.

### Example: Access a top-level field

Given the following example document:

```json
{
  "user": "alice"
}
```

> You can access `user` using:

Reviewer: ```suggestion
You can access `user` as follows:
```

```json
"field": "ctx.user"
```

### Example: Access nested fields

Given the following example document:

```json
{
  "user": {
    "name": "alice"
  }
}
```

> You can access `user.name` using:

Reviewer: ```suggestion
You can access `user.name` as follows:
```

```json
"field": "ctx.user.name"
```

## Accessing fields in the source

> To access a field in the document `_source`, refer to fields by their name:

Reviewer: ```suggestion
To access a field in the document `_source`, refer to fields by their names:
```

```json
{
  "set": {
    "field": "environment",
    "value": "production"
  }
}
```

> Or explicitly use `_source`:

Reviewer: ```suggestion
Alternatively, you can explicitly use `_source`:
```

```json
{
  "set": {
    "field": "_source.environment",
    "value": "production"
  }
}
```

## Accessing metadata fields

> You can read or write metadata fields such as:

Reviewer: ```suggestion
You can read or write to metadata fields such as:
```

- `_index`
- `_type`
- `_id`
- `_routing`

### Example: Set `_routing` dynamically

```json
{
  "set": {
    "field": "_routing",
    "value": "{% raw %}{{region}}{% endraw %}"
  }
}
```

Using `{% raw %}{{_id}}{% endraw %}` is not supported when document IDs are auto-generated.
{: .note}

## Accessing ingest metadata fields

> The `_ingest.timestamp` field represents when the ingest node received the document. To persist this timestamp:

Reviewer: ```suggestion
The `_ingest.timestamp` field represents the time when the ingest node received the document. To persist this timestamp, use the `set` processor:
```

```json
{
  "set": {
    "field": "received_at",
    "value": "{% raw %}{{_ingest.timestamp}}{% endraw %}"
  }
}
```

## Using `ctx` in Mustache templates

> Use Mustache templates to insert field values into processor settings. Use triple curly braces for unescaped field values.

Reviewer: ```suggestion
Use Mustache templates to insert field values into processor settings. Use triple curly braces (`{{{` and `}}}`) for unescaped field values.
```

### Example: Combine source fields

> ```json

Reviewer: Add an intro sentence for this example.

{
  "set": {
    "field": "log_label",
    "value": "{% raw %}{{{app}}}_{{{env}}}{% endraw %}"
  }
}
```

> ### Example: Dynamic greeting using `set` processor

Reviewer: ```suggestion
### Example: Generating a dynamic greeting using the set processor
```

> If the document has `"user": "alice"`, use the following syntax to produce the result `"greeting": "Hello, alice!"`.

Reviewer: ```suggestion
If a document's `user` field is set to `alice`, use the following syntax to produce the result `"greeting": "Hello, alice!"`.
```

```json
{
  "set": {
    "field": "greeting",
    "value": "Hello, {% raw %}{{{user}}}{% endraw %}!"
  }
}
```

## Dynamic field names

You can use a field's value as the name of a new field.

```json
{
  "set": {
    "field": "{% raw %}{{service}}{% endraw %}",
    "value": "{% raw %}{{code}}{% endraw %}"
  }
}
```

> ## Example: Route to dynamic index based on status

Reviewer: ```suggestion
## Example: Routing to a dynamic index based on status
```

> ```json

Reviewer: Please add an intro sentence for this request.

{
  "set": {
    "field": "_index",
    "value": "{% raw %}{{status}}{% endraw %}-events"
  }
}
```

## Using `ctx` in script processors

Use the `script` processor for advanced transformations.

> ### Example: Add a field only if another is missing

Reviewer: ```suggestion
### Example: Adding a field only if another is missing
```

> 

Reviewer: Please add an intro sentence for this request.

{: .note}

## Accessing ingest metadata fields

The `_ingest.timestamp` field represents the time when the ingest node received the document. To persist this timestamp, use the `set` processor:

```json
{
  "set": {
    "field": "received_at",
    "value": "{% raw %}{{_ingest.timestamp}}{% endraw %}"
  }
}
```

## Using `ctx` in Mustache templates

Use Mustache templates to insert field values into processor settings. Use triple curly braces (`{{{` and `}}}`) for unescaped field values.

### Example: Combining source fields

The following processor configuration combines the `app` and `env` fields, separated by an underscore (_), and stores the result in the `log_label` field:

```json
{
  "set": {
    "field": "log_label",
    "value": "{% raw %}{{{app}}}_{{{env}}}{% endraw %}"
  }
}
```

### Example: Generating a dynamic greeting using the set processor

If a document's `user` field is set to `alice`, use the following syntax to produce the result `"greeting": "Hello, alice!"`.

```json
{
  "set": {
    "field": "greeting",
    "value": "Hello, {% raw %}{{{user}}}{% endraw %}!"
  }
}
```

## Dynamic field names

> You can use a field's value as the name of a new field.

Reviewer 2: ```suggestion
You can use a field's value as the name of a new field:
```

```json
{
  "set": {
    "field": "{% raw %}{{service}}{% endraw %}",
    "value": "{% raw %}{{code}}{% endraw %}"
  }
}
```

## Example: Routing to a dynamic index based on status

> The following processor configuration sets the target index dynamically by appending `-events` to the value of the `status` field.

Reviewer 2: ```suggestion
The following processor configuration sets the target index dynamically by appending `-events` to the value of the `status` field:
```

```json
{
  "set": {
    "field": "_index",
    "value": "{% raw %}{{status}}{% endraw %}-events"
  }
}
```

> ## Using `ctx` in script processors

Reviewer 2: ```suggestion
## Using `ctx` in the `script` processor
```

Use the `script` processor for advanced transformations.

### Example: Adding a field only if another is missing

The following processor adds the `error_message` field with the value "none" only if the field is missing from the document:

```json
{
  "script": {
    "lang": "painless",
    "source": "if (ctx.error_message == null) { ctx.error_message = 'none'; }"
  }
}
```

### Example: Copying a value from one field to another

The following processor copies the value from the `timestamp` field into a new field called `event_time`:

```json
{
  "script": {
    "lang": "painless",
    "source": "ctx.event_time = ctx.timestamp;"
  }
}
```

> ## A complete pipeline example

Reviewer 2: ```suggestion
## Example of a complete pipeline
```

> The following example defines a complete ingest pipeline that sets a tagline using `source` field, extracts the `year` from the `date` field, and records the document’s ingest timestamp in the `received_at` field:

Reviewer 2: ```suggestion
The following example defines a complete ingest pipeline that sets a tagline using the `source` field, extracts the `year` from the `date` field, and records the document’s ingest timestamp in the `received_at` field:
```

```json
PUT _ingest/pipeline/example-pipeline
{
  "description": "Sets tags, log label, and defaults error message",
  "processors": [
    {
      "set": {
        "field": "tagline",
        "value": "{% raw %}{{{user.first}}} from {{{department}}}{% endraw %}"
      }
    },
    {
      "script": {
        "lang": "painless",
        "source": "ctx.year = ctx.date.substring(0, 4);"
      }
    },
    {
      "set": {
        "field": "received_at",
        "value": "{% raw %}{{_ingest.timestamp}}{% endraw %}"
      }
    }
  ]
}
```

To test the pipeline, use the following request:

```json
POST _ingest/pipeline/example-pipeline/_simulate
{
  "docs": [
    {
      "_source": {
        "user": {
          "first": "Liam"
        },
        "department": "Engineering",
        "date": "2024-12-03T14:05:00Z"
      }
    }
  ]
}
```

> The response shows the enriched document after processing, including the newly added `tagline`, extracted year, and the `received_at` timestamp generated by the ingest pipeline:

Reviewer 2: Should `year` be in code font here, as above?
